### PR TITLE
gh-104146: Argument clinic: Remove dead code flagged by mypy's `truthy-bool` check

### PR DESCRIPTION
**Pull request by AlexWaygood**
And enable that check in our mypy config.

* Issue: gh-104146

### CODE REVIEW

File: Tools/clinic/clinic.py [hunk 1]
```diff
@@ -4428,7 +4426,7 @@ class DSLParser:
     keyword_only: bool
     positional_only: bool
     group: int
-    parameter_state: int
+    parameter_state: ParamState
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:4429]: This isn't related to the `truthy-bool` check, but it's a tiny typing-related fix that doesn't feel like it deserves its own PR. Mypy doesn't flag it (since it's not technically incorrect, it could just be slightly more specific), but I noticed pyright flagged it when I was editing `clinic.py` in VSCode the other day.

File: Tools/clinic/clinic.py [hunk 2]
```diff
@@ -886,9 +886,7 @@ def output_templates(
         converters = [p.converter for p in parameters]
 
         has_option_groups = parameters and (parameters[0].group or parameters[-1].group)
-        default_return_converter = (not f.return_converter or
-            f.return_converter.type == 'PyObject *')
-
+        default_return_converter = f.return_converter.type == 'PyObject *'
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:889]: `f.return_converter` is always an instance of `CReturnConverter`, which is always truthy:

https://github.com/python/cpython/blob/ac7605ed197e8b2336d44c8ac8aeae6faa90a768/Tools/clinic/clinic.py#L2540C23-L2540C39

File: Tools/clinic/clinic.py [hunk 3]
```diff
@@ -4790,8 +4788,6 @@ def state_modulename_name(self, line: str) -> None:
         if not return_converter:
             return_converter = CReturnConverter()
 
-        if not module:
-            fail("Undefined module used in declaration of {full_name.strip()!r}.")
```
Comment (AlexWaygood) [inline Tools/clinic/clinic.py:4794]: `module` is always either an instance of `Module` or an instance of `Clinic`, and both are always truthy

**Review by erlend-aasland — APPROVED** [review]
Nice, thanks!
